keps/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,227 @@
+- `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology: topology-2`
+
+The leader pod will first be scheduled on a placement group. Once a worker pod is created, it will follow the leader to the placement group. 
+Afterward, it will be scheduled into a nodepool that has other pods with the same subgroup index. So, the placement will look something like this
+
+![kep-115 scheduling](https://github.com/kubernetes-sigs/lws/assets/86417275/33c49e9f-84da-4ac9-8e2b-973b532a0756)
+
+### Implementation
+- Two new annotations will be added
+  - `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology`
+  - `leaderworkerset.sigs.k8s.io/subgroup-size` 
+- Three new labels will be added,
+  - `leaderworkerset.sigs.k8s.io/subgroup-index = worker-index/subGroupSize`
+    - Tracks which subgroup the pod is part of 
+  - `leaderworkerset.sigs.k8s.io/subgroup-worker-index = worker-index%subGroupSize`
+    - index/identity of a pod inside the pod's subgroup
+  - `leaderworkerset.sigs.k8s.io/subgroup-key` 
+    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index
+
+To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 
+
+### TPU Environment Variable Injection
+Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_WORKER_HOSTNAMES` will only be a list of the pods in the same subgroup.
+
+LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two vm hosts each. Since the leader doesn’t request TPU resources, it won't take full ownership of the TPU vm, allowing 
+a worker that does request TPU resources to run in the same vm. Therefore, there will be four workers + the leader, meaning that one of the workers will have a worker index of four, a
+as shown in the picture below. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.
+
+![Leader doesn't request TPU resources](https://github.com/kubernetes-sigs/lws/assets/86417275/3a1292ac-7221-4324-acda-ff33702838f9)
+
+If the leader does not request TPU resources, then the labels will have the following values
+`leaderworkerset.sigs.k8s.io/subgroup-index = (workerIndex - 1) / subGroupSize`
+`leaderworkerset.sigs.k8s.io/subgroup-worker-index = (workerIndex - 1) % subGroupSize`
```

Comment:
We can't change the semantics of subgroup-index and subgroup-worker-index based on tpu requests, what we change is the value we set to TPU_WORKER_ID.

Reply:
As suggested above, I don't think we need subgroup-worker-index for now.

Reply:
ditto